p3/_utils.py

```diff
@@ -30,3 +30,21 @@ def _require_numeric(df, columns):
         except Exception:
             msg = "Column '%s' must contain only numeric values."
             raise TypeError(msg % (column))
+
+
+def _sort_by_app_order(df, app_order):
```

Comment:
This function seems really convoluted and I might just be missing the reason why because I don't have a clear mental model for what the dataframe might look like, could you give a really small example of the input/output desired?

I've also never sorted by strings this way before, so it might just be what's required

Reply:
You're right to question it.  I struggle to make sense of the pandas interface at the best of times, and so a lot of the solutions I come up with are not necessarily the best ones.

If we initially have a DataFrame like this:
| Application | FOM |
|-------------|-----|
| A           | 0.5 |
| B           | 0.4 |
| B           | 0.9 |
| A           | 0.7 |

...then `df["application"].unique()` returns `["A", "B"]`.

If we filter the original DataFrame such that only the best FOMs remain, then the DataFrame looks like this:

| Application | FOM |
|-------------|-----|
| B           | 0.9 |
| A           | 0.7 |

...and `df["application"].unique()` returns `["B", "A"]`.

So what this function does is accept the second (post-filtering) DataFrame and a list denoting the desired application order (i.e., `["A", "B"]`), and then sorts the DataFrame such that the results appear in the same order:

| Application | FOM |
|-------------|-----|
| A           | 0.7 |
| B           | 0.9 |

Note that this isn't the same as sorting the application names.  What we're trying to do is ensure that we don't alter the order in which applications appear in the DataFrame, in case that order is meaningful to the user.

As a concrete example... Imagine that the results in the DataFrame are all sorted by collection date, and that a user is appending new results to the DataFrame as they're collected.  Without this convoluted sorting, adding a new (better) result to the bottom of the DataFrame could change the order in which applications appear, resulting in a different legend the next time the graph is plotted.

Does that make sense?

Reply:
Inspired by our conversation about "doing one thing well" on another pull request, I think I'm coming around to the view that #22 probably was a mistake.  I'll sketch out an alternative fix and open that as another pull request, so we can compare the two possible behaviors.